File: libs/io/src/main/java/org/kop/libs/io/net/HttpClient.java
package org.kop.libs.io.net;

import jdk.jfr.Experimental;
import lombok.Getter;
import lombok.SneakyThrows;
import lombok.val;
import okhttp3.*;
import org.jetbrains.annotations.NotNull;

import java.io.File;

@Experimental
@Getter
public class HttpClient {
    private final OkHttpClient client;

    private static final MediaType JSON
            = MediaType.get("application/json; charset=utf-8");

    private int cacheSize = 10 * 1024 * 1024;

    private Cache cache = new Cache(new File(System.getProperty("java.io.tmpdir")), this.cacheSize);

    public HttpClient(OkHttpClient.@NotNull Builder builder) {
        this.client = builder.cache(cache).build();
    }

    public HttpClient(OkHttpClient.@NotNull Builder builder, int cacheSize) {
        this.cacheSize = cacheSize;
        this.client = builder.cache(cache).build();
    }

    public HttpClient(OkHttpClient.@NotNull Builder builder, Cache cache) {
        this.client = builder.cache(cache).build();
        this.cache = cache;
    }

    @SneakyThrows
    public Response get(Request.@NotNull Builder reqBuilder) {
        val req = reqBuilder.get().build();
        return exec(req);
    }

    public Response get(String url) {
        val req = new Request.Builder().url(url).get().build();
        return exec(req);
    }

    @SneakyThrows
    public Response exec(Request req) {
        try (val resp = client.newCall(req).execute()) {
            return resp;
        }
    }

    public Response postJson(Request.@NotNull Builder reqBuilder, String json) {
        RequestBody body = RequestBody.create(json, JSON);
        val req = reqBuilder.post(body).build();
        return exec(req);
    }
}

File: libs/io/src/main/kotlin/org/toolkit4j/libs/io/files/HttpClient.kt
package org.toolkit4j.libs.io.files

import java.io.File
import jdk.jfr.Experimental
import lombok.Getter
import lombok.SneakyThrows
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaTypeOrNull

@Experimental
@Getter
class HttpClient {
  private val mediaType: MediaType? = "application/json; charset=utf-8".toMediaTypeOrNull()
  private val client: OkHttpClient
  private var cacheSize = 10 * 1024 * 1024
  private var cache = Cache(File(System.getProperty("java.io.tmpdir")), cacheSize.toLong())

  constructor(builder: OkHttpClient.Builder) {
    client = builder.cache(cache).build()
  }

  constructor(builder: OkHttpClient.Builder, cacheSize: Int) {
    this.cacheSize = cacheSize
    client = builder.cache(cache).build()
  }

  constructor(builder: OkHttpClient.Builder, cache: Cache) {
    client = builder.cache(cache).build()
    this.cache = cache
  }

  @SneakyThrows
  operator fun get(reqBuilder: Request.Builder): Response {
    val req: Request = reqBuilder.get().build()
    return exec(req)
  }

  operator fun get(url: String?): Response {
    val req: Request? = url?.let { Request.Builder().url(it).get().build() }
    return exec(req)
  }

  @SneakyThrows
  fun exec(req: Request?): Response {
    client.newCall(req!!).execute().use { resp ->
      return resp
    }
  }

  fun postJson(reqBuilder: Request.Builder, json: String?): Response {
    val body: RequestBody? = json?.let { RequestBody.create(mediaType, it) }
    val req: Request? = body?.let { reqBuilder.post(it).build() }
    return exec(req)
  }
}
